File: src/main/java/uk/yermak/audiobookconverter/Conversion.java
package uk.yermak.audiobookconverter;

import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.collections.ObservableList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static uk.yermak.audiobookconverter.ProgressStatus.*;

/**
 * Created by Yermak on 06-Feb-18.
 */
public class Conversion {
    final static Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

    private final static ExecutorService executorService = Executors.newCachedThreadPool();
    private SimpleObjectProperty<ProgressStatus> status = new SimpleObjectProperty<>(this, "status", READY);

    private OutputParameters outputParameters;
    private String outputDestination;
    private Convertable convertable;
    private AudioBookInfo bookInfo;
    private List<ArtWork> posters;

    public List<MediaInfo> getMedia() {
        return getConverable().getMedia();
    }


    public void start(Convertable convertable, String outputDestination, Refreshable refreshable, OutputParameters outputParameters, AudioBookInfo bookInfo, ObservableList<ArtWork> posters) {
        this.convertable = convertable;
        this.outputDestination = outputDestination;
        this.outputParameters = outputParameters;
        this.bookInfo = bookInfo;
        this.posters = new ArrayList<>(posters);

        Executors.newSingleThreadExecutor().execute(refreshable);

        Map<String, ProgressCallback> progressCallbacks = new HashMap<>();

        convertable.getMedia().stream().map(m -> (m.getFileName() + "-" + m.getDuration())).forEach(key -> progressCallbacks.put(key, new ProgressCallback(key, refreshable)));


        progressCallbacks.put("output", new ProgressCallback("output", refreshable));

        ConversionStrategy conversionStrategy = new ParallelConversionStrategy(this, progressCallbacks);

        executorService.execute(conversionStrategy);
        status.set(IN_PROGRESS);
    }


    public void addStatusChangeListener(ChangeListener<ProgressStatus> listener) {
        status.addListener(listener);
    }

    public void pause() {
        if (status.get().equals(IN_PROGRESS)) {
            status.set(PAUSED);
        }
    }

    public void stop() {
        if (!status.get().equals(FINISHED)) {
            status.set(CANCELLED);
        }
    }

    public ProgressStatus getStatus() {
        return status.get();
    }


    public void finished() {
        status.set(FINISHED);
    }

    public void error(String message) {
        status.set(ERROR);
    }

    public void resume() {
        if (status.get().equals(PAUSED)) {
            status.set(IN_PROGRESS);
        }
    }

    public OutputParameters getOutputParameters() {
        return outputParameters;
    }


    public String getOutputDestination() {
        return outputDestination;
    }

    public Convertable getConverable() {
        return convertable;
    }

    public AudioBookInfo getBookInfo() {
        return bookInfo;
    }

    public List<ArtWork> getPosters() {
        return posters;
    }

}

File: src/main/scala/uk/yermak/audiobookconverter/Conversion.scala
package uk.yermak.audiobookconverter

import java.util.concurrent.Executors

import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.collections.{FXCollections, ObservableList}
import uk.yermak.audiobookconverter.ConversionMode.ConversionMode
import uk.yermak.audiobookconverter.ProgressStatus._

/**
  * Created by rYermak on 06-Feb-18.
  */
object Conversion {
  private val executorService = Executors.newCachedThreadPool
}

class Conversion {
  private val media = FXCollections.observableArrayList[MediaInfo]
  private val mode = new SimpleObjectProperty[ConversionMode](ConversionMode.PARALLEL)
  private val status = new SimpleObjectProperty[ProgressStatus](this, "status", READY)
  private var bookInfo: AudioBookInfo = _
  private var outputParameters: OutputParameters = _
  private var outputDestination: String = _

  def setMode(mode: ConversionMode) = this.mode.set(mode)

  def setBookInfo(bookInfo: AudioBookInfo) = this.bookInfo = bookInfo

  def getMedia: ObservableList[MediaInfo] = media

  def getMode: ConversionMode = mode.get

  def getBookInfo: AudioBookInfo = bookInfo

  def start(outputDestination: String, refreshable: Refreshable) = {
    setOutputDestination(outputDestination)
    Executors.newSingleThreadExecutor.execute(refreshable)
    val progressCallbacks: Map[String, ProgressCallback] = Map()
    media.forEach((mediaInfo: MediaInfo) => progressCallbacks + (mediaInfo.getFileName -> new ProgressCallback(mediaInfo.getFileName, refreshable)))
    progressCallbacks + ("output" -> new ProgressCallback("output", refreshable))
    val conversionStrategy: ConversionStrategy = ConversionMode.createConvertionStrategy(this, progressCallbacks)
    Conversion.executorService.execute(conversionStrategy)
    status.set(IN_PROGRESS)
  }

  def addStatusChangeListener(listener: ChangeListener[ProgressStatus]) = status.addListener(listener)

  def pause() = if (status.get == IN_PROGRESS) status.set(PAUSED)

  def stop() = if (!(status.get == FINISHED)) status.set(CANCELLED)

  def getStatus = status.get

  def finished() = status.set(FINISHED)

  def error(message: String) = status.set(ERROR)

  def resume() = if (status.get == PAUSED) status.set(IN_PROGRESS)

  def addModeChangeListener(listener: ChangeListener[ConversionMode]) = mode.addListener(listener)

  def setOutputParameters(params: OutputParameters) = outputParameters = params

  def getOutputParameters = outputParameters

  def getPosters = bookInfo.getPosters

  def getOutputDestination = outputDestination

  def setOutputDestination(outputDestination: String) = this.outputDestination = outputDestination
}